app/src/middleware/apiTracker.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * @module ApiTracker
+ *
+ * Log statistics for CDOGS Requests.
+ *
+ * @see morgan
+ *
+ * @exports initializeApiTracker
+ */
+const moment = require('moment');
+const morgan = require('morgan');
+
+// this will suppress a console warning about moment deprecating a default fallback on non ISO/RFC2822 date formats
+// we will just force it to use the new Date constructor.
+moment.createFromInputFallback = function (config) {
+  config._d = new Date(config._i);
+};
+
+const docGenUrl = '/api/v1/docGen';
+const trackerUrls = [docGenUrl];
+
+// add in any token (custom or morgan built-in) we want to the format, then morgan can parse out later
+// status and response-time is a built-in morgan token
+const apiTrackerFormat = ':operation :azp :timestamp :contextKeyCount :contentFileType :contentEncodingType :contentSize :outputFileType :res[content-length] :status :response-time';
+
+const apiTracker = async (req, res, next) => {
+
+  if (trackerUrls.includes(req.url)) {
+    req._timestamp = moment.utc().valueOf();
+    req._operation = req.url === docGenUrl ? 'DOCGEN' : 'Unknown';
+
+    /*
+    When/If we need to parse data out of the response, we would do it here...
+    const defaultEnd = res.end;
+    const chunks = [];
+    res.end = (...restArgs) => {
+        try {
+            if (restArgs[0]) {
+                chunks.push(Buffer.from(restArgs[0]));
+            }
+            const body = Buffer.concat(chunks).toString('utf8');
+            const obj = JSON.parse(body);
+
+        } catch (err) {
+            log.error('mailApiTracker', err);
+        }
+        defaultEnd.apply(res, restArgs);
+    };
+    */
+  }
+  next();
+};
+
+const initializeApiTracker = (app) => {
+
+  // register token parser functions.
+  // this one would depend on authorizedParty middleware being loaded
+  morgan.token('azp', (req) => {
+    return req.authorizedParty ? req.authorizedParty : '-';
+  });
+
+  morgan.token('operation', (req) => {
+    return req._operation ? req._operation : '-';
+  });
+
+  morgan.token('timestamp', (req) => {
+    return req._timestamp ? req._timestamp : '-';
+  });
+
+  morgan.token('contextKeyCount', (req) => {
+    function countKeys(source) {
```

Comment:
Think it's worth moving this count to a util function? Would be an easy unit test then as well.
Or is it so hyper-specific to this logging need that there's no point...?

Reply:
Pretty specific to this case... will have to wait and see if it's even what Matthew wants.  But, if it is, it should probably go into utils and get tested.

Reply:
Yeah should wait and see I suppose

Reply:
The lambda functions only have one argument so they don't need the parentheses around req.
Yea I agree with the wait and see -> don't have enough requirements set yet to know where it should end up just yet.